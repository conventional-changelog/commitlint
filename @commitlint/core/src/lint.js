import {entries} from 'lodash';
import isIgnored from './library/is-ignored';
import parse from './library/parse';
import implementations from './rules';

export default async (message, rules = {}, opts = {}) => {
	// Found a wildcard match, skip
	if (isIgnored(message)) {
		return {
			valid: true,
			errors: [],
			warnings: []
		};
	}

	// Parse the commit message
	const parsed = await parse(message, undefined, opts.parserOpts);

	// Validate against all rules
	const results = entries(rules)
		.filter(entry => {
			const [, [level]] = entry;
			return level > 0;
		})
		.map(entry => {
			const [name, config] = entry;
			const [level, when, value] = config;

			// Level 0 rules are ignored
			if (level === 0) {
				return null;
			}

			const rule = implementations[name];
			const [valid, message] = rule(parsed, when, value);

			return {
				level,
				valid,
				name,
				message
			};
		})
		.filter(Boolean);

	const errors = results.filter(result => result.level === 2 && !result.valid);
	const warnings = results.filter(
		result => result.level === 1 && !result.valid
	);

	const valid = errors.length === 0;

	return {
		valid,
		errors,
		warnings
	};
};
