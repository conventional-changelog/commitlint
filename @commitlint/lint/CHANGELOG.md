# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [14.1.0](https://github.com/conventional-changelog/commitlint/compare/v14.0.0...v14.1.0) (2021-11-01)

**Note:** Version bump only for package @commitlint/lint





# [14.0.0](https://github.com/conventional-changelog/commitlint/compare/v13.2.1...v14.0.0) (2021-10-26)

**Note:** Version bump only for package @commitlint/lint





# [13.2.0](https://github.com/conventional-changelog/commitlint/compare/v13.1.0...v13.2.0) (2021-09-28)

**Note:** Version bump only for package @commitlint/lint





# [13.1.0](https://github.com/conventional-changelog/commitlint/compare/v13.0.0...v13.1.0) (2021-07-24)

**Note:** Version bump only for package @commitlint/lint





# [13.0.0](https://github.com/conventional-changelog/commitlint/compare/v12.1.4...v13.0.0) (2021-05-24)


* chore!: remove node 10 support (#2596) ([4db4ba1](https://github.com/conventional-changelog/commitlint/commit/4db4ba1b0b312410a0f62100a93a80c246a6c410)), closes [#2596](https://github.com/conventional-changelog/commitlint/issues/2596)


### BREAKING CHANGES

* minimum node version is 12





## [12.1.3](https://github.com/conventional-changelog/commitlint/compare/v12.1.2...v12.1.3) (2021-05-12)

**Note:** Version bump only for package @commitlint/lint





## [12.1.2](https://github.com/conventional-changelog/commitlint/compare/v12.1.1...v12.1.2) (2021-04-29)

**Note:** Version bump only for package @commitlint/lint





## [12.1.1](https://github.com/conventional-changelog/commitlint/compare/v12.1.0...v12.1.1) (2021-04-02)

**Note:** Version bump only for package @commitlint/lint





# [12.1.0](https://github.com/conventional-changelog/commitlint/compare/v12.0.1...v12.1.0) (2021-03-06)

**Note:** Version bump only for package @commitlint/lint





## [12.0.1](https://github.com/conventional-changelog/commitlint/compare/v12.0.0...v12.0.1) (2021-02-23)

**Note:** Version bump only for package @commitlint/lint





# [12.0.0](https://github.com/conventional-changelog/commitlint/compare/v11.0.0...v12.0.0) (2021-01-18)

**Note:** Version bump only for package @commitlint/lint





# [11.0.0](https://github.com/conventional-changelog/commitlint/compare/v10.0.0...v11.0.0) (2020-09-05)

**Note:** Version bump only for package @commitlint/lint





# [10.0.0](https://github.com/conventional-changelog/commitlint/compare/v9.1.2...v10.0.0) (2020-08-16)


* refactor!: drop support for node 8 (#1999) ([751f39f](https://github.com/conventional-changelog/commitlint/commit/751f39f284ef232574a176c3c11b1982ee544166)), closes [#1999](https://github.com/conventional-changelog/commitlint/issues/1999)


### BREAKING CHANGES

* remove node 8 from circle-ci checks

also remove node 13 because we do not support experimental versions

* docs: update node v10 to latest LTS 10 version

Co-authored-by: Cedric van Putten <me@bycedric.com>

Co-authored-by: Cedric van Putten <me@bycedric.com>





## [9.1.2](https://github.com/conventional-changelog/commitlint/compare/v9.1.1...v9.1.2) (2020-07-13)

**Note:** Version bump only for package @commitlint/lint





## [9.1.1](https://github.com/conventional-changelog/commitlint/compare/v9.1.0...v9.1.1) (2020-06-30)

**Note:** Version bump only for package @commitlint/lint





# [9.1.0](https://github.com/conventional-changelog/commitlint/compare/v9.0.1...v9.1.0) (2020-06-21)

**Note:** Version bump only for package @commitlint/lint





## [9.0.1](https://github.com/conventional-changelog/commitlint/compare/v9.0.0...v9.0.1) (2020-05-26)

**Note:** Version bump only for package @commitlint/lint





# [9.0.0](https://github.com/conventional-changelog/commitlint/compare/v8.3.5...v9.0.0) (2020-05-21)


### Bug Fixes

* [#840](https://github.com/conventional-changelog/commitlint/issues/840) add caret to lodash versions ([#843](https://github.com/conventional-changelog/commitlint/issues/843)) ([ffc0bac](https://github.com/conventional-changelog/commitlint/commit/ffc0bac26993acb2ab6a8fa51065f93c92b0d644))
* ignore empty commit messages [#615](https://github.com/conventional-changelog/commitlint/issues/615) ([#676](https://github.com/conventional-changelog/commitlint/issues/676)) ([c3eb1a7](https://github.com/conventional-changelog/commitlint/commit/c3eb1a76e08213d7ce1f200e35f8d5d6de18982a))


### Features

* add async promise based rules methods into lint ([#976](https://github.com/conventional-changelog/commitlint/issues/976)) ([4443062](https://github.com/conventional-changelog/commitlint/commit/444306249b8a3d04524538f61edca8f6cc10d75f))





## [8.3.5](https://github.com/conventional-changelog/commitlint/compare/v8.3.4...v8.3.5) (2020-01-15)

**Note:** Version bump only for package @commitlint/lint





## [8.3.4](https://github.com/conventional-changelog/commitlint/compare/v8.3.3...v8.3.4) (2020-01-03)

**Note:** Version bump only for package @commitlint/lint





# [8.3.0](https://github.com/conventional-changelog/commitlint/compare/v8.2.0...v8.3.0) (2019-10-16)

**Note:** Version bump only for package @commitlint/lint





# [8.2.0](https://github.com/conventional-changelog/commitlint/compare/v8.1.0...v8.2.0) (2019-09-16)

**Note:** Version bump only for package @commitlint/lint





# [8.1.0](https://github.com/conventional-changelog/commitlint/compare/v8.0.0...v8.1.0) (2019-07-15)

**Note:** Version bump only for package @commitlint/lint





# [7.6.0](https://github.com/conventional-changelog/commitlint/compare/v7.5.2...v7.6.0) (2019-05-06)


### Features

* adds support for plugins ([#228](https://github.com/conventional-changelog/commitlint/issues/228)) ([#588](https://github.com/conventional-changelog/commitlint/issues/588)) ([cea4564](https://github.com/conventional-changelog/commitlint/commit/cea4564))
* config based is-ignored overrides ([#595](https://github.com/conventional-changelog/commitlint/issues/595)) ([2434d71](https://github.com/conventional-changelog/commitlint/commit/2434d71))





## [7.5.2](https://github.com/conventional-changelog/commitlint/compare/v7.5.1...v7.5.2) (2019-02-11)

**Note:** Version bump only for package @commitlint/lint





## [7.5.1](https://github.com/conventional-changelog/commitlint/compare/v7.5.0...v7.5.1) (2019-02-09)

**Note:** Version bump only for package @commitlint/lint





<a name="7.5.0"></a>
# [7.5.0](https://github.com/conventional-changelog/commitlint/compare/v7.4.0...v7.5.0) (2019-01-31)


### Bug Fixes

* all broken website references ([#564](https://github.com/conventional-changelog/commitlint/issues/564)) ([82eeb5a](https://github.com/conventional-changelog/commitlint/commit/82eeb5a))
* replace all repository references with conventional changelog ([#561](https://github.com/conventional-changelog/commitlint/issues/561)) ([6c3afcd](https://github.com/conventional-changelog/commitlint/commit/6c3afcd))
* replace all website references with conventional changelog ([#563](https://github.com/conventional-changelog/commitlint/issues/563)) ([6b86fb1](https://github.com/conventional-changelog/commitlint/commit/6b86fb1))




<a name="7.4.0"></a>
# [7.4.0](https://github.com/conventional-changelog/commitlint/compare/v7.3.2...v7.4.0) (2019-01-27)




**Note:** Version bump only for package @commitlint/lint

<a name="7.3.2"></a>
## [7.3.2](https://github.com/conventional-changelog/commitlint/compare/v7.3.1...v7.3.2) (2019-01-15)


### Bug Fixes

* move loadsh to dependencies ([1bb66a4](https://github.com/conventional-changelog/commitlint/commit/1bb66a4))




<a name="7.3.1"></a>
## [7.3.1](https://github.com/conventional-changelog/commitlint/compare/v7.3.0...v7.3.1) (2019-01-11)




**Note:** Version bump only for package @commitlint/lint

<a name="7.3.0"></a>
# [7.3.0](https://github.com/conventional-changelog/commitlint/compare/v7.2.1...v7.3.0) (2019-01-11)


### Bug Fixes

* update to security-patched dependency versions ([97c033b](https://github.com/conventional-changelog/commitlint/commit/97c033b))




<a name="7.2.1"></a>
## [7.2.1](https://github.com/conventional-changelog/commitlint/compare/v7.2.0...v7.2.1) (2018-10-11)




**Note:** Version bump only for package @commitlint/lint

<a name="7.2.0"></a>
# [7.2.0](https://github.com/conventional-changelog/commitlint/compare/v7.1.2...v7.2.0) (2018-10-05)




**Note:** Version bump only for package @commitlint/lint

<a name="7.1.2"></a>
## [7.1.2](https://github.com/conventional-changelog/commitlint/compare/v7.1.1...v7.1.2) (2018-09-04)




**Note:** Version bump only for package @commitlint/lint

<a name="7.1.0"></a>
# [7.1.0](https://github.com/conventional-changelog/commitlint/compare/v7.0.1...v7.1.0) (2018-08-29)




**Note:** Version bump only for package @commitlint/lint

<a name="6.2.0"></a>
# [6.2.0](https://github.com/conventional-changelog/commitlint/compare/v6.1.3...v6.2.0) (2018-05-01)


### Features

* print commit message when the message is invalid ([86c34f1](https://github.com/conventional-changelog/commitlint/commit/86c34f1)), closes [#222](https://github.com/conventional-changelog/commitlint/issues/222)




<a name="6.1.1"></a>
## [6.1.1](https://github.com/conventional-changelog/commitlint/compare/v6.1.0...v6.1.1) (2018-02-22)




**Note:** Version bump only for package @commitlint/lint
