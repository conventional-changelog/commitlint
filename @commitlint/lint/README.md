> Lint a string against commitlint rules

# @commitlint/lint

## Getting started

```shell
npm install --save @commitlint/lint
```

## Example

```js
const lint = require('@commitlint/lint');

lint('foo: bar', {'type-enum': [1, 'always', ['foo']]})
  .then(report => console.log(report));
  // => { valid: true, errors: [], warnings: [] }

lint('foo: bar', {'type-enum': [1, 'always', ['bar']]})
  .then(report => console.log(report));
  /* =>
    { valid: true,
      errors: [],
      warnings:
      [ { level: 1,
          valid: false,
          name: 'type-enum',
          message: 'type must be one of [bar]' } ] }
  */
```

Consult [docs/api](https://conventional-changelog.github.io/commitlint/#/reference-api) for comprehensive documentation.
