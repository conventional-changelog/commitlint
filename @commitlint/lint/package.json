{
  "name": "@commitlint/lint",
  "version": "7.0.0",
  "description": "Lint a string against commitlint rules",
  "main": "lib/index.js",
  "files": [
    "lib/"
  ],
  "scripts": {
    "build": "cross-env NODE_ENV=production babel src --out-dir lib  --source-maps",
    "clean": "npx rimraf lib",
    "deps": "dep-check",
    "pkg": "pkg-check --skip-import",
    "lint": "xo",
    "start": "concurrently \"ava -c 4 --verbose --watch\" \"yarn run watch\"",
    "test": "ava -c 4 --verbose",
    "watch": "babel src --out-dir lib --watch --source-maps"
  },
  "ava": {
    "files": [
      "src/**/*.test.js",
      "!lib/**/*"
    ],
    "source": [
      "src/**/*.js",
      "!lib/**/*"
    ],
    "babel": "inherit",
    "require": [
      "babel-register"
    ]
  },
  "babel": {
    "presets": [
      "babel-preset-commitlint"
    ]
  },
  "xo": false,
  "engines": {
    "node": ">=4"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/marionebl/commitlint.git"
  },
  "bugs": {
    "url": "https://github.com/marionebl/commitlint/issues"
  },
  "homepage": "https://github.com/marionebl/commitlint#readme",
  "keywords": [
    "conventional-changelog",
    "commitlint",
    "library",
    "core"
  ],
  "author": {
    "name": "Mario Nebl",
    "email": "hello@herebecode.com"
  },
  "license": "MIT",
  "devDependencies": {
    "@commitlint/test": "7.0.0",
    "@commitlint/utils": "^7.0.0",
    "ava": "0.22.0",
    "babel-cli": "6.26.0",
    "babel-preset-commitlint": "^7.0.0",
    "babel-register": "6.26.0",
    "concurrently": "3.5.1",
    "cross-env": "5.1.1",
    "execa": "0.9.0",
    "globby": "8.0.1",
    "lodash.includes": "4.3.0",
    "rimraf": "2.6.1",
    "xo": "0.20.3"
  },
  "dependencies": {
    "@commitlint/is-ignored": "^7.0.0",
    "@commitlint/parse": "^7.0.0",
    "@commitlint/rules": "^7.0.0",
    "babel-runtime": "^6.23.0",
    "lodash.topairs": "4.3.0"
  }
}
