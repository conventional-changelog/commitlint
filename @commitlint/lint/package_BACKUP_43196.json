{
  "name": "@commitlint/lint",
  "version": "8.3.4",
  "description": "Lint a string against commitlint rules",
  "main": "lib/index.js",
  "files": [
    "lib/"
  ],
  "scripts": {
    "build": "cross-env NODE_ENV=production babel src --out-dir lib  --source-maps",
    "deps": "dep-check",
    "pkg": "pkg-check --skip-import",
    "start": "concurrently \"ava -c 4 --verbose --watch\" \"yarn run watch\"",
    "test": "ava -c 4 --verbose",
    "watch": "babel src --out-dir lib --watch --source-maps"
  },
  "ava": {
    "files": [
      "src/**/*.test.js",
      "!lib/**/*"
    ],
    "source": [
      "src/**/*.js",
      "!lib/**/*"
    ],
    "babel": {
      "testOptions": {
        "presets": [
          "babel-preset-commitlint"
        ]
      }
    },
    "require": [
      "@babel/register"
    ]
  },
  "babel": {
    "presets": [
      "babel-preset-commitlint"
    ]
  },
  "engines": {
    "node": ">=4"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/conventional-changelog/commitlint.git"
  },
  "bugs": {
    "url": "https://github.com/conventional-changelog/commitlint/issues"
  },
  "homepage": "https://github.com/conventional-changelog/commitlint#readme",
  "keywords": [
    "conventional-changelog",
    "commitlint",
    "library",
    "core"
  ],
  "author": {
    "name": "Mario Nebl",
    "email": "hello@herebecode.com"
  },
  "license": "MIT",
  "devDependencies": {
    "@babel/cli": "^7.7.7",
    "@babel/core": "^7.7.7",
    "@babel/register": "^7.7.7",
    "@commitlint/test": "8.2.0",
<<<<<<< HEAD
    "@commitlint/utils": "^8.3.0",
    "ava": "2.4.0",
=======
    "@commitlint/utils": "^8.3.4",
    "ava": "0.22.0",
    "babel-cli": "6.26.0",
>>>>>>> upstream/master
    "babel-preset-commitlint": "^8.2.0",
    "concurrently": "3.6.1",
    "cross-env": "6.0.3",
    "execa": "0.11.0",
    "globby": "10.0.1",
    "proxyquire": "2.1.3"
  },
  "dependencies": {
<<<<<<< HEAD
    "@commitlint/is-ignored": "^8.3.2",
    "@commitlint/parse": "^8.3.2",
    "@commitlint/rules": "^8.3.2",
=======
    "@commitlint/is-ignored": "^8.3.4",
    "@commitlint/parse": "^8.3.4",
    "@commitlint/rules": "^8.3.4",
    "babel-runtime": "^6.23.0",
>>>>>>> upstream/master
    "lodash": "4.17.15"
  }
}
