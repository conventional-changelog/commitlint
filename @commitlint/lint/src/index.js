import util from 'util';
import isIgnored from '@commitlint/is-ignored';
import parse from '@commitlint/parse';
import implementations from '@commitlint/rules';
import entries from 'lodash.topairs';

const buildCommitMesage = ({header, body, footer}) => {
	let message = header;

	message = body ? `${message}\n\n${body}` : message;
	message = footer ? `${message}\n\n${footer}` : message;

	return message;
};

export default async (message, rules = {}, opts = {}) => {
	// Found a wildcard match, skip
	if (isIgnored(message)) {
		return {
			valid: true,
			errors: [],
			warnings: [],
			input: message
		};
	}

	// Parse the commit message
	const parsed = await parse(message, undefined, opts.parserOpts);

	// Find invalid rules configs
	const missing = Object.keys(rules).filter(
		name => typeof implementations[name] !== 'function'
	);

	if (missing.length > 0) {
		const names = Object.keys(implementations);
		throw new RangeError(
			`Found invalid rule names: ${missing.join(
				', '
			)}. Supported rule names are: ${names.join(', ')}`
		);
	}

	const invalid = entries(rules)
		.map(([name, config]) => {
			if (!Array.isArray(config)) {
				return new Error(
					`config for rule ${name} must be array, received ${util.inspect(
						config
					)} of type ${typeof config}`
				);
			}

			const [level, when] = config;

			if (typeof level !== 'number' || isNaN(level)) {
				return new Error(
					`level for rule ${name} must be number, received ${util.inspect(
						level
					)} of type ${typeof level}`
				);
			}

			if (level === 0 && config.length === 1) {
				return null;
			}

			if (config.length !== 2 && config.length !== 3) {
				return new Error(
					`config for rule ${name} must be 2 or 3 items long, received ${util.inspect(
						config
					)} of length ${config.length}`
				);
			}

			if (level < 0 || level > 2) {
				return new RangeError(
					`level for rule ${name} must be between 0 and 2, received ${util.inspect(
						level
					)}`
				);
			}

			if (typeof when !== 'string') {
				return new Error(
					`condition for rule ${name} must be string, received ${util.inspect(
						when
					)} of type ${typeof when}`
				);
			}

			if (when !== 'never' && when !== 'always') {
				return new Error(
					`condition for rule ${name} must be "always" or "never", received ${util.inspect(
						when
					)}`
				);
			}

			return null;
		})
		.filter(item => item instanceof Error);

	if (invalid.length > 0) {
		throw new Error(invalid.map(i => i.message).join('\n'));
	}

	// Validate against all rules
	const results = entries(rules)
		.filter(entry => {
			const [, [level]] = entry;
			return level > 0;
		})
		.map(entry => {
			const [name, config] = entry;
			const [level, when, value] = config;

			// Level 0 rules are ignored
			if (level === 0) {
				return null;
			}

			const rule = implementations[name];

			const [valid, message] = rule(parsed, when, value);

			return {
				level,
				valid,
				name,
				message
			};
		})
		.filter(Boolean);

	const errors = results.filter(result => result.level === 2 && !result.valid);
	const warnings = results.filter(
		result => result.level === 1 && !result.valid
	);

	const valid = errors.length === 0;

	return {
		valid,
		errors,
		warnings,
		input: buildCommitMesage(parsed)
	};
};
