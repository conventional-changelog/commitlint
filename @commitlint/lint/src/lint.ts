import util from 'util';
import isIgnored from '@commitlint/is-ignored';
import parse from '@commitlint/parse';
import defaultRules from '@commitlint/rules';
import {buildCommitMesage} from './commit-message';
import {
	LintOptions,
	LintOutcome,
	LintRuleOutcome,
	Rule,
	RuleConfigSeverity,
	BaseRule,
	RuleType,
	QualifiedRules,
} from '@commitlint/types';

export default async function lint(
	message: string,
	rawRulesConfig?: QualifiedRules,
	rawOpts?: LintOptions
): Promise<LintOutcome> {
	const opts = rawOpts
		? rawOpts
		: {defaultIgnores: undefined, ignores: undefined};
	const rulesConfig = rawRulesConfig || {};

	// Found a wildcard match, skip
	if (
		isIgnored(message, {defaults: opts.defaultIgnores, ignores: opts.ignores})
	) {
		return {
			valid: true,
			errors: [],
			warnings: [],
			input: message,
		};
	}

	// Parse the commit message
	const parsed =
		message === ''
			? {header: null, body: null, footer: null}
			: await parse(message, undefined, opts.parserOpts);

	if (
		parsed.header === null &&
		parsed.body === null &&
		parsed.footer === null
	) {
		// Commit is empty, skip
		return {
			valid: true,
			errors: [],
			warnings: [],
			input: message,
		};
	}

	const allRules: Map<string, BaseRule<never, RuleType>> = new Map(
		Object.entries(defaultRules)
	);

	if (opts.plugins) {
		Object.values(opts.plugins).forEach((plugin) => {
			if (plugin.rules) {
				Object.keys(plugin.rules).forEach((ruleKey) =>
					allRules.set(ruleKey, plugin.rules[ruleKey])
				);
			}
		});
	}

	// Find invalid rules configs
	const missing = Object.keys(rulesConfig).filter(
		(name) => typeof allRules.get(name) !== 'function'
	);

	if (missing.length > 0) {
		const names = [...allRules.keys()];
		throw new RangeError(
			`Found invalid rule names: ${missing.join(
				', '
			)}. Supported rule names are: ${names.join(', ')}`
		);
	}

	const invalid = Object.entries(rulesConfig)
		.map(([name, config]) => {
			if (!Array.isArray(config)) {
				return new Error(
					`config for rule ${name} must be array, received ${util.inspect(
						config
					)} of type ${typeof config}`
				);
			}

			const [level] = config;

			if (level === RuleConfigSeverity.Disabled && config.length === 1) {
				return null;
			}

			const [, when] = config;

			if (typeof level !== 'number' || isNaN(level)) {
				return new Error(
					`level for rule ${name} must be number, received ${util.inspect(
						level
					)} of type ${typeof level}`
				);
			}

			if (config.length < 2 || config.length > 3) {
				return new Error(
					`config for rule ${name} must be 2 or 3 items long, received ${util.inspect(
						config
					)} of length ${config.length}`
				);
			}

			if (level < 0 || level > 2) {
				return new RangeError(
					`level for rule ${name} must be between 0 and 2, received ${util.inspect(
						level
					)}`
				);
			}

			if (typeof when !== 'string') {
				return new Error(
					`condition for rule ${name} must be string, received ${util.inspect(
						when
					)} of type ${typeof when}`
				);
			}

			if (when !== 'never' && when !== 'always') {
				return new Error(
					`condition for rule ${name} must be "always" or "never", received ${util.inspect(
						when
					)}`
				);
			}

			return null;
		})
		.filter((item): item is Error => item instanceof Error);

	if (invalid.length > 0) {
		throw new Error(invalid.map((i) => i.message).join('\n'));
	}

	// Validate against all rules
	const pendingResults = Object.entries(rulesConfig)
		// Level 0 rules are ignored
		.filter(([, config]) => !!config && config.length && config[0] > 0)
		.map(async (entry) => {
			const [name, config] = entry;
			const [level, when, value] = config!; //

			const rule = allRules.get(name);

			if (!rule) {
				throw new Error(`Could not find rule implementation for ${name}`);
			}

			const executableRule = rule as Rule<unknown>;
			const [valid, message] = await executableRule(parsed, when, value);

			return {
				level,
				valid,
				name,
				message,
			};
		});

	const results = (await Promise.all(pendingResults)).filter(
		(result): result is LintRuleOutcome => result !== null
	);

	const errors = results.filter(
		(result) => result.level === 2 && !result.valid
	);
	const warnings = results.filter(
		(result) => result.level === 1 && !result.valid
	);

	const valid = errors.length === 0;

	return {
		valid,
		errors,
		warnings,
		input: buildCommitMesage(parsed),
	};
}
