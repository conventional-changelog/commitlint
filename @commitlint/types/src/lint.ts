import {IsIgnoredOptions} from './is-ignored';
import {PluginRecords} from './load';
import {ParserOptions} from './parse';
import {RuleConfigSeverity, RuleConfigTuple} from './rules';

export type LintRuleConfig = Record<
	string,
	| Readonly<[RuleConfigSeverity.Disabled]>
	| RuleConfigTuple<void>
	| RuleConfigTuple<unknown>
>;

export interface LintOptions {
	/** If it should ignore the default commit messages (defaults to `true`) */
	defaultIgnores?: IsIgnoredOptions['defaults'];
	/** Additional commits to ignore, defined by ignore matchers  */
	ignores?: IsIgnoredOptions['ignores'];
	/** The parser configuration to use when linting the commit */
	parserOpts?: ParserOptions;

	plugins?: PluginRecords;
	helpUrl?: string;
}

export interface LintOutcome {
	/** The linted commit, as string */
	input: string;
	/** If the linted commit is considered valid */
	valid: boolean;
	/** All errors, per rule, for the commit */
	errors: LintRuleOutcome[];
	/** All warnings, per rule, for the commit */
	warnings: LintRuleOutcome[];
}

export interface LintRuleOutcome {
	/** If the commit is considered valid for the rule */
	valid: boolean;
	/** The "severity" of the rule (1 = warning, 2 = error) */
	level: RuleConfigSeverity;
	/** The name of the rule */
	name: string;
	/** The message returned from the rule, if invalid */
	message: string;
}
