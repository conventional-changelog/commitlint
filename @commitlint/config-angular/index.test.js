import lint from '@commitlint/lint';
import {rules, parserPreset} from '.';

const lintMessage = async (message) => {
	const parserOpts = parserPreset.parserOpts;
	const m = message.replace(/^\s+/, '').trim();
	const result = await lint(m, rules, {parserOpts});

	if (result.errors.length > 1) {
		throw new Error(
			'Commit test should only have one error message to validate against'
		);
	}

	if (result.warnings.length > 1) {
		throw new Error(
			'Commit test should only have one warning message to validate against'
		);
	}

	return result;
};

test('a valid commit message', async () => {
	const result = await lintMessage('test: a valid angular commit');
	expect(result.valid).toBe(true);
	expect(result.errors).toStrictEqual([]);
	expect(result.warnings).toStrictEqual([]);
});

test('a valid message with a scope', async () => {
	const result = await lintMessage(
		'test(scope): a valid angular commit with a scope'
	);
	expect(result.valid).toBe(true);
	expect(result.errors).toStrictEqual([]);
	expect(result.warnings).toStrictEqual([]);
});

test('a valid multi line commit', async () => {
	const result = await lintMessage(
		`test(scope): a valid angular commit with a scope

     Some content in the body`
	);
	expect(result.valid).toBe(true);
	expect(result.errors).toStrictEqual([]);
	expect(result.warnings).toStrictEqual([]);
});

test('a leading blank line after header', async () => {
	const result = await lintMessage(
		`test(scope): a valid angular commit with a scope
     Some content in the body`
	);

	expect(result.valid).toBe(true);
	expect(result.errors).toStrictEqual([]);
	expect(result.warnings[0].message).toBe('body must have leading blank line');
});

test('an invalid scope', async () => {
	const result = await lintMessage(`no: no is not not an invalid commit type`);

	expect(result.valid).toBe(false);
	expect(result.errors[0].message).toBe(
		'type must be one of [build, ci, docs, feat, fix, perf, refactor, revert, style, test]'
	);
	expect(result.warnings).toStrictEqual([]);
});

test('a long header', async () => {
	const result = await lintMessage(
		`test: that its an error when there is ia realllllllllllllllllllllly long header`
	);

	expect(result.valid).toBe(false);
	expect(result.errors[0].message).toBe(
		'header must not be longer than 72 characters, current length is 79'
	);
	expect(result.warnings).toStrictEqual([]);
});

test('message header with ! in it', async () => {
	const result = await lintMessage(`test!: with a breaking change in the type`);

	expect(result.valid).toBe(false);
	expect(result.errors[0].message).toBe(
		'subject must not have an exclamation mark in the subject to identify a breaking change'
	);
	expect(result.warnings).toStrictEqual([]);
});

test('message header with ! in it and a scope', async () => {
	const result = await lintMessage(
		`test(scope)!: with a breaking change in the type`
	);

	expect(result.valid).toBe(false);
	expect(result.errors[0].message).toBe(
		'subject must not have an exclamation mark in the subject to identify a breaking change'
	);
	expect(result.warnings).toStrictEqual([]);
});
