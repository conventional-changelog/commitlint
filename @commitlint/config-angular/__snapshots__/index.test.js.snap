// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rules body-leading-blank empty 1`] = `
Array [
  false,
  "body must have leading blank line",
]
`;

exports[`rules body-leading-blank filled 1`] = `
Array [
  true,
]
`;

exports[`rules body-leading-blank simple 1`] = `
Array [
  true,
]
`;

exports[`rules footer-leading-blank empty 1`] = `
Array [
  true,
]
`;

exports[`rules footer-leading-blank filled 1`] = `
Array [
  false,
  "footer must have leading blank line",
]
`;

exports[`rules footer-leading-blank simple 1`] = `
Array [
  true,
]
`;

exports[`rules header-max-length empty 1`] = `
Array [
  false,
  "header must not be longer than 2,always,72 characters, current length is 13",
]
`;

exports[`rules header-max-length filled 1`] = `
Array [
  false,
  "header must not be longer than 2,always,72 characters, current length is 13",
]
`;

exports[`rules header-max-length simple 1`] = `
Array [
  false,
  "header must not be longer than 2,always,72 characters, current length is 13",
]
`;

exports[`rules scope-case empty 1`] = `
Array [
  true,
]
`;

exports[`rules scope-case filled 1`] = `
Array [
  true,
]
`;

exports[`rules scope-case simple 1`] = `
Array [
  true,
]
`;

exports[`rules subject-case empty 1`] = `
Array [
  true,
  "subject must be , never, ",
]
`;

exports[`rules subject-case filled 1`] = `
Array [
  true,
  "subject must be , never, ",
]
`;

exports[`rules subject-case simple 1`] = `
Array [
  true,
  "subject must be , never, ",
]
`;

exports[`rules subject-empty empty 1`] = `
Array [
  false,
  "subject must be empty",
]
`;

exports[`rules subject-empty filled 1`] = `
Array [
  false,
  "subject must be empty",
]
`;

exports[`rules subject-empty simple 1`] = `
Array [
  false,
  "subject must be empty",
]
`;

exports[`rules subject-full-stop empty 1`] = `
Array [
  false,
  "subject must end with full stop",
]
`;

exports[`rules subject-full-stop filled 1`] = `
Array [
  false,
  "subject must end with full stop",
]
`;

exports[`rules subject-full-stop simple 1`] = `
Array [
  false,
  "subject must end with full stop",
]
`;

exports[`rules type-case empty 1`] = `
Array [
  true,
  "type must be , always, lower-case",
]
`;

exports[`rules type-case filled 1`] = `
Array [
  true,
  "type must be , always, lower-case",
]
`;

exports[`rules type-case simple 1`] = `
Array [
  true,
  "type must be , always, lower-case",
]
`;

exports[`rules type-empty empty 1`] = `
Array [
  false,
  "type must be empty",
]
`;

exports[`rules type-empty filled 1`] = `
Array [
  false,
  "type must be empty",
]
`;

exports[`rules type-empty simple 1`] = `
Array [
  false,
  "type must be empty",
]
`;

exports[`rules type-enum empty 1`] = `
Array [
  false,
  "type must be one of [2, always, build,ci,docs,feat,fix,perf,refactor,revert,style,test]",
]
`;

exports[`rules type-enum filled 1`] = `
Array [
  false,
  "type must be one of [2, always, build,ci,docs,feat,fix,perf,refactor,revert,style,test]",
]
`;

exports[`rules type-enum simple 1`] = `
Array [
  false,
  "type must be one of [2, always, build,ci,docs,feat,fix,perf,refactor,revert,style,test]",
]
`;
